### ISSUE TITLE
Rule Change: give an option to `max-params` to omit `this` no matter its type

### ISSUE THREAD

**Issue by GerkinDev** [eslint/eslint#20107]
### What rule do you want to change?

max-params

### What change do you want to make?

Generate fewer warnings

### How do you think the change should be implemented?

A new option

### Example code

```js
// With `{max: 2}`
function doSomething(this: MyType, param1: unknown, param2: unknown): unknown {}
```

### What does the rule currently do for this code?

Currently, the rule fails with the code above, despite `this` not being a parameter. 

### What will the rule do after it's changed?

No warning is generated if the new option (for example, `countThis`, similar to [`countVoidThis`](https://eslint.org/docs/latest/rules/max-params#countvoidthis-typescript-only)) is set to `false`. For backward compatibility, it can default to `true`

### Participation

- [x] I am willing to submit a pull request to implement this change.

### Additional comments

Initially opened via https://github.com/typescript-eslint/typescript-eslint/issues/11574 but they plan on deprecating the rule since it has been imported here.

**Comment by mdjermanovic** [comment]
I'm not sure why only `this: void` was excluded from counting parameters in the typescript-eslint extension rule as there wasn't much discussion in the original issue (https://github.com/typescript-eslint/typescript-eslint/issues/7538).

I think it would make sense to not count `this` as a parameter regardless of the type, because it technically isn't a parameter, and I agree with the proposal to put this behavior behind an option that defaults to the current behavior. We should just think about how the new option would work together with the existing option `countVoidThis`.

@eslint/eslint-team thoughts about accepting this proposal, and the design?

**Comment by fasttime** [comment]
> I think it would make sense to not count `this` as a parameter regardless of the type, because it technically isn't a parameter, and I agree with the proposal to put this behavior behind an option that defaults to the current behavior. We should just think about how the new option would work together with the existing option `countVoidThis`.
> 
> [@eslint/eslint-team](https://github.com/orgs/eslint/teams/eslint-team) thoughts about accepting this proposal, and the design?

We could perhaps deprecate `countVoidThis` in favor of a new option `countThis` that would accept one of three values, for example `"always"` (same as `countVoidThis: true`), `"never"`, and `"except-void"` (default). The new option would be mutually exclusive with `countVoidThis`.

**Comment by Revanwjy-alt** [comment]
main disini berhadiah menarik

jo777.help

**Comment by JoshuaKGoldberg** [comment]
I'm in favor of @fasttime's proposal. 👍

**Comment by nzakas** [comment]
I'm in favor of @fasttime's proposal. @mdjermanovic what do you think?

**Comment by mdjermanovic** [comment]
@fasttime proposal sounds good to me too. Marking as accepted, PR is welcome.

**Comment by loontrix** [comment]
@mdjermanovic Would it be okay if I send a PR?

**Comment by mdjermanovic** [comment]
@luentrix the issue is already assigned to @GerkinDev.

@GerkinDev would you like to send a PR for this change?

**Comment by GerkinDev** [comment]
Yes I will, I was waiting for a confirmation about how, which I just had
